### ISSUE TITLE
Supporting OpenGL

### ISSUE THREAD

**Issue by lite20** [NodeOS/NodeOS#270]
I've been looking into enabling OpenGL as this is needed for the workspace flavour (aka graphic edition of NodeOS). I understand that we'll need to compile mesa, and package it into an npm module which being completely honest, really scares me. Nonetheless I'm going to give a shot at compiling mesa for the framebuffer as is accomplished [here](https://blog.visucore.com/2013/2/25/building-mesa-without-x11). This thread is to monitor my progress creating mesa module!

**Comment by lite20** [comment]
I'm thinking Mesa's dependencies might make it a pain to work with. None the less, pressing on to see where I can get to.

**Comment by lite20** [comment]
Maybe we could consider first starting with http://bellard.org/TinyGL/ ? It's small size makes it relatively quick and I quote from it's landing page:

>  TinyGL is a lot faster than Mesa

It would be a nice start for us and we could slowly grow toward the monolith mesa. Why do we need mesa? Mesa implements the full OpenGL standard, and by missing parts, we limit what programs can be run. I suggest we make it a param on the build scripts of the graphic edition to either use mesa or a smaller renderer.

**Comment by lite20** [comment]
Update:

So I read somewhere that TinyGL was made in 2002, making me question if it's recent enough to be used for anything at all. I will experiment with it in a bit but at the moment I'm compiling mesa. I'll have you know that mesa takes a good million years to compile, but doesn't use much ram. It's size might make it impractical but I'll have a better say on that if it finishes compiling successfully. I'm working to create an npm package that could auto compile it should we choose to use mesa in the graphic editions (therefor, I'm writing down all the commands I use and steps I follow). Mesa has been compiling for over an hour now and show's no signs of stopping... I regret not using my other core.

**Comment by lite20** [comment]
**Update**
So after 2+ hours, I was successful at compiling mesa for DRM. This is most likely what we'd use (I believe) if we were to use mesa on NodeOS so seeing that I can compile it is a good sign. I'm not quite sure what I just did however as mesa doesn't appear have created any new files (at a first glance). I'll need to investigate what was compiled, and where it is. Once I find this, it'll be a matter of hacking this build into the NodeOS build scripts to try and see if I can get it to work in NodeOS. I'll have the scripts also set everything up and it won't be a clean set up but for a first proof-of-concept prototype that should be okay.

This is full mesa so we can expect great things from it XD. I've investigated QEMU's OpenGL emulation and I believe I know what to do to get it working. Hopefully we'll have a prototype to play with soon. AS it is now 1 am (the build took longer than foreshadowed), I'm now off to sleep. Will report any new finds tomorrow!

**Comment by piranna** [comment]
> and package it into an npm module which being completely honest, really scares me

Don't think too much on the npm package yet, let's make it work first. One step at a time ;-)

Why would we need Mesa? For node-webgl, for example? Would it need it noGUI for something? Does this compilation has hardware acceleration, or only just a software raster?

**Comment by piranna** [comment]
> Does this compilation has hardware acceleration, or only just a software raster?

```sh
./configure --enable-gles2 --enable-gles1 --disable-glx --enable-egl --enable-gallium-egl \
     --disable-dri --with-gallium-drivers=swrast --with-egl-platforms=fbdev
```

Software raster, but with egl support. Ouh yeah B-) Having this compiling is a good start, it would be cool to have `node-webgl` to work with this :-) After that we could start enabling DRI and other things to try to have hardware acceleration using just the build in kernel drivers :-) That lead me to think... what can we do with them? Add another options on the build process to select the devices that will be included on the kernel, or should we still left to use the default kernel config?

**Comment by Coretool** [comment]
When we've got OpenGL running, I can finally use CEF for noGUI ... Life's getting easier :D

**Comment by lite20** [comment]
> Why would we need Mesa? For node-webgl, for example? Would it need it noGUI for something? Does this compilation has hardware acceleration, or only just a software raster?

Mesa would be an excellent OpenGL implementation for us since it can take over where hardware lacks. Mesa will use hardware acceleration where available, or software where it cannot. Right now the precise configuration I'm using enables egl coupled with soft raster as @piranna guessed. I figured it would be a good start.


> ./configure --enable-gles2 --enable-gles1 --disable-glx --enable-egl --enable-gallium-egl \
     --enable-dri --with-gallium-drivers=swrast --with-egl-platforms=drm

**Comment by lite20** [comment]
@piranna Mesa will be or gateway to supporting any 3D rendering be it for webgl, virtual reality, or quite literally anything using gl. We open the gates of 3D to our graphic users (which is an immense world) making NodeOS just that more powerful.

**Warning**
I do heavily suggest the Mesa be an optional include even for the graphic distribution as it is no small project. It's feasible and something we could actually manage to pack by the look of things, but iz definitely an extra that  select group of NodeOS users would be using. (WebGL game developers, or gamers, etc)

**Comment by lite20** [comment]
> When we've got OpenGL running, I can finally use CEF for noGUI ... Life's getting easier :D

I've got your back!

**Comment by piranna** [comment]
> Warning
> I do heavily suggest the Mesa be an optional include even for the graphic
distribution as it is no small project. It's feasible and something we
could actually manage to pack by the look of things, but iz definitely an
extra that select group of NodeOS users would be using. (WebGL game
developers, or gamers, etc)

I think so, that's why I suggest don't work too much on the npm package
while we experiment with it, to see if we can compile it statically onto
node-webgl or CEF or whatever. We need to know before how it works and how
to integrate it before take a decission.

**Comment by lite20** [comment]
@piranna I'm running a second compile (this time taking advantage of all my cores) and am beginning to get an idea of how this all works. OpenGL systems dynamic. Programs using OpenGL link up with the dynamic library available on any system and they all use the same names, and interfaces so an OpenGL program could have no idea what raster it's using, whether it be a proprietary driver, or an soft-render.

To my understanding, in order to use mesa we would need to place the generated .so files in the 'lib' folder (that's created when mesa is compiled) in it's generic home so installed OpenGL programs can use it. (Be it CEF, or practically anything using OpenGL, as users may chose to swap out their renderer, who knows?). The generic home for OpenGL libraries is `/usr/include/GL` as noted [here](http://stackoverflow.com/questions/2182790/where-are-the-opengl-libraries-get-stored-on-ubuntu-i-need-this-to-mention-in). I'm however curious where the `/usr/bin/` gl files come from, as I don't see them in the compiled mesa project. I'll have to look deeper and maybe scavenge the net to grab their sources and additionally compile them.

**Comment by piranna** [comment]
> @piranna I'm running a second compile (this time taking advantage of all
my cores) and am beginning to get an idea of how this all works. OpenGL
systems dynamic. Programs using OpenGL link up with the dynamic library
available on any system and they all use the same names, and interfaces so
an OpenGL program could have no idea what raster it's using, whether it be
a proprietary driver, or an soft-render.

Mostly yes, that's how the fragments connect between them in the jigsaw.

> To my understanding, in order to use mesa we would need to place the
generated .so files in the 'lib' folder (that's created when mesa is
compiled) in it's generic home so installed OpenGL programs can use it. (Be
it CEF, or practically anything using OpenGL, as users may chose to swap
out their renderer, who knows?).

Here we have two options, make each user to have its own copy so they can
easily change version, or since it's related to the system hardware common
to all the users, place it on the initram of the workstation flavor similar
to how its done with libfuse. In any case, users would be able to override
it anyway...

> The generic home for OpenGL libraries is /usr/include/GL as noted here.

I don't think this will be needed in a production system so it should bot
be in the initram, but it would be to build it.

> I'm however curious where the /usr/bin/ gl files come from, as I don't
see them in the compiled mesa project. I'll have to look deeper and maybe
scavenge the net to grab their sources and additionally compile them.

I think we'll not need them, just only libGL.so, no more.

**Comment by lite20** [comment]
> I think we'll not need them, just only libGL.so, no more.

Okay, awesome!

> Mostly yes, that's how the fragments connect between them in the jigsaw.

Excellent. Second build done and I have all these prebuilt libs to play with. What's the easiest way to inject them into NodeOS? wget?

> Here we have two options, make each user to have its own copy so they can
easily change version, or since it's related to the system hardware common
to all the users, place it on the initram of the workstation flavor similar
to how its done with libfuse. In any case, users would be able to override
it anyway...

I suggest a common installation, to avoid it's massive size being replicated and eating space pointlessly.

**Comment by piranna** [comment]
>> Mostly yes, that's how the fragments connect between them in the jigsaw.
>
> Excellent. Second build done and I have all these prebuilt libs to play
with. What's the easiest way to inject them into NodeOS? wget?

Copy them somewhere nodeos-initramfs project inside your NodeOS
node_modules and change its scripts/build by hand to include them the same
way it's done with libfuse. Maybe you could also add them directly to its
build folder...

**Comment by lite20** [comment]
> Copy them somewhere nodeos-initramfs project inside your NodeOS
node_modules and change its scripts/build by hand to include them the same
way it's done with libfuse. Maybe you could also add them directly to its
build folder...

Okay awesome, thanks! I'll give it a shot.

>  > The generic home for OpenGL libraries is /usr/include/GL as noted here.
> I don't think this will be needed in a production system so it should bot
be in the initram, but it would be to build it.

So it doesn't matter where we place the .so files?

**Comment by piranna** [comment]
>>> The generic home for OpenGL libraries is /usr/include/GL as noted here.
>>> I don't think this will be needed in a production system so it should
bot
>>> be in the initram, but it would be to build it.
>
> So it doesn't matter where we place the .so files?

As far as they end on runtime on the /lib folder it's fine.

**Comment by joshgarde** [comment]
Any ideas on what display server you're going to use?

**Comment by joshgarde** [comment]
And as a follow up, window manager?

**Comment by Coretool** [comment]
AFAIK we won't use any display server. The apps will all be like 'normal websites'. We basically use a browser engine, currently we're using webkit but I want to swicht as soon as possible to the chormium embedded framework. Th current project can be found at [the noGui repo](http://www.github.com/coretool/noGui)

**Comment by joshgarde** [comment]
I was thinking of porting over Chromium OS' ash window system once I was done with my Mac ports of the build tools.

**Comment by joshgarde** [comment]
Not for the official builds, just for my own lols.

**Comment by Coretool** [comment]
Yeah, I've looked at ash... I like it but I don't know if that matches our philosophy (@piranna, what do you think ?)

**Comment by joshgarde** [comment]
I find it the most lightweight solution to implementing a display server + window manager even though all it can display is Chromium.

**Comment by Coretool** [comment]
Yeah... but does it use X ?

**Comment by joshgarde** [comment]
Google switched over to the Freon display stack last year: https://www.phoronix.com/scan.php?page=news_item&px=Chrome-OS-Freon-Graphics

**Comment by Coretool** [comment]
Do you know where the ash (or aura shell ?) source is ?

**Comment by Coretool** [comment]
Yeah... saw that project ^^

**Comment by Coretool** [comment]
Now I know what I'll do in my holidays ^^ Thanks for links ;)

**Comment by joshgarde** [comment]
I'd be happy to help out

**Comment by piranna** [comment]
> Yeah, I've looked at ash... I like it but I don't know if that matches
our philosophy (@piranna, what do you think ?)

As far as we can use it without the chrome-only APIs...

**Comment by joshgarde** [comment]
Woo!

**Comment by SpaceboyRoss01** [comment]
I'm going to write my own desktop/gui/window manager package for NodeOS.

**Comment by Coretool** [comment]
Wouldn't it be more efficient if we all worked on the same project ?

**Comment by joshgarde** [comment]
I believe it would be, so who's project should we focus on?

**Comment by Coretool** [comment]
Well I started some time ago with [noGui](https://www.github.com/coretool/noGui). I think we will record our progress there. The most developed attempt is using QTwebkit and would only need some logging and some changes to the network... 

However, I would love to try to make ash/aura/freon work (with nodeos) !

**Comment by joshgarde** [comment]
I'm going to get started on ash on https://github.com/joshgarde/aurora

**Comment by joshgarde** [comment]
I forgot to say in my original comment that I'd be happy to contribute code to noGui as well.

**Comment by Coretool** [comment]
Ok... I'll add you to noGui... Will add me to aurora too ?

**Comment by joshgarde** [comment]
Done :D

**Comment by joshgarde** [comment]
Just checked out the chromium source (15GB), started building it (3 hours ago) on a new VPS I had to purchase to meet the [system requirements](https://chromium.googlesource.com/chromium/src/+/master/docs/linux_build_instructions.md#System-requirements), and realized that porting ash probably isn't worth the work.

So I'm now changing my original plans for aurora - which might be an extension of noGui for window management relying simply on EGL for display.

**Comment by joshgarde** [comment]
And now back to the main topic of this issue: How's mesa support going?

**Comment by piranna** [comment]
> So I'm now changing my original plans for aurora - which might become an extension of noGui for window management relying simply on EGL for display.

The idea is to do everything in vanilla Javascript, if you can run it in a web browser you can run it on noGUI, so no extensions here. The only extension I would allow is the X-FRAME to disable the iframe spoofing and being able to navigate to Google or Github or Facebook from inside noGUI, and this because there are already some extensions for regular browsers that enable some whitelist sites to disable it too.

**Comment by joshgarde** [comment]
I wasn't talking about a browser extension. I'll have to explain my idea when I'm more alert (stayed up all night).

I don't feel like having a web browser within a web browser is a very good idea since one unstable website would cause the entire desktop to crash.

**Comment by piranna** [comment]
> But I don't feel like having a web browser within a web browser is a very good idea since one unstable website would cause the entire desktop to crash.

That's why we plan to use iframes for each app, to isolate them ones from others.

**Comment by joshgarde** [comment]
A quick test in Chrome doesn't quite support this:

isolation-test.html:
```
<html>
<head>
<title></title>
</head>
<body>

<iframe src="embeded-crash.html"></iframe>
<iframe src="https://www.duckduckgo.com/"></iframe>

</body>
</html>
```

embeded-crash.html:
```
<html>
<head>
<script>
txt = "a";
while(1)
{
    txt = txt += "a";    //add as much as the browser can handle
}
</script>
<body>

</body>
</html>
```


Also, anything in an iframe can access the main document though `window.parent`

**Comment by piranna** [comment]
> Also, anything in an iframe can access the main document though window.parent

If they are from different domains no, they don't have access.

**Comment by lite20** [comment]
> And now back to the main topic of this issue: How's mesa support going?

 I haven't gotten a chance to sit with it for the last two days but I got a lot of work hacking away at NodeOS and can now inject files right into a NodeOS workspace so I'm going to attempt to jet Mesa right into my NodeOS set up and see if I can get it going. I'll also try jetting gcc (probably g++) in there too while I'm at it to test how NodeOS handles it.

**Comment by piranna** [comment]
How have you done it?

**Comment by lite20** [comment]
> How have you done it?

Yeah I did get Mesa on. G++ was however taking a billions years to compile and I couldn't finish. I'm now working on creating a simple binary using OpenGL to inject into NodeOS and test if Mesa works

**Comment by piranna** [comment]
That's cool :-) In case of doubts, take a look how it's done on Linux From
Scratch.

**Comment by lite20** [comment]
@piranna good suggestion! I've bee super busy but hopefully I'll get to play with this on Monday.

**Comment by piranna** [comment]
Any update on this?

**Comment by joshgarde** [comment]
I'd be happy to try my hand at this project if that's alright with @lite20

**Comment by piranna** [comment]
> I'd be happy to try my hand at this project if that's alright with @lite20

If you are interested give it a try, maybe later would you both join efforts :-)

**Comment by lite20** [comment]
> I'd be happy to try my hand at this project if that's alright with @lite20

Fine by me! We can use this thread to communicate. I'm working to get back to work w/o a computer (got a VPS), but I'm sure you can work faster than I with yours. I'll dump any finds here for you to work with!

**Comment by lite20** [comment]
Just wondering if anyone has taken a chip at this? I haven't been able to work much on NodeOS as my linux pc has been out of commission but I've got it going again. I'll mess with this a bit more and see what I can get done.

Did you make any strides @joshgarde ?

**Comment by joshgarde** [comment]
@lite20 I've been waiting for NodeOS to get to a stable build before trying to hack together any sort of GUI solution.

**Comment by joshgarde** [comment]
I do however want to still continue down the Chromium OS route and see where that can take us.

**Comment by lite20** [comment]
@joshgarde yeah that isn't a bad idea. I've been messing with that right now to see if I can get anything going.

**Comment by piranna** [comment]
There's no problem on you guys working on the UI also while NodeOS is still unstable, as I told other times by design using a regular web browser and connect to a remote NodeOS instance is a perfectly valid use case, and later move to a local web renderer would be trivial this way.

**Comment by adminy** [comment]
Guys this issue is Support for OpenGL, not make your own window manager, if you wanna do that go download an open source linux copy which is very good at doing those things. (Sorry if it sounds harsh)

Can this be implemented as an npm package with an accessible API?

**Comment by SpaceboyRoss01** [comment]
It can be implemented but there would need to be some sort of API that accesses a framebuffer device file and somehow gets an OpenGL context for it.

**Comment by piranna** [comment]
Main problem is lack of time and/or sponsors to work on that :-/

**Comment by adminy** [comment]
Can’t it be ported with emscripten or the likes? Time is precious to everyone indeed. I had a look at TinyGL, and although it works for 3D graphics it's missing one essential part of webGL, the use of program. So methods like `createProgram` and `useProgram` are not available in TinyGL when the environment is not a browser. There is [headless-gl](https://github.com/stackgl/headless-gl) But it's extremely tied to `ANGLE` and OpenGL. I'm surprised that there is not a single Emulator out there that does just OpenGL emulation. The two things that are interesting to look at are Microsoft Edge Implementation of [WebGL](https://github.com/MicrosoftEdge/WebGL) and Internet Explorer's [WebGL-Plugin](https://github.com/iewebgl/iewebgl). There are lots of bindings available that make OpenGL into WebGL API.
